src/app/clusters/ecosystem-information-server/ecosystem-information-server.cpp

```diff
@@ -0,0 +1,375 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include "ecosystem-information-server.h"
+
+#include <app/AttributeAccessInterface.h>
+#include <app/AttributeAccessInterfaceRegistry.h>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace EcosystemInformation {
+namespace {
+
+constexpr size_t kDeviceNameMaxSize             = 64;
+constexpr size_t kUniqueLocationIdMaxSize       = 64;
+constexpr size_t kUniqueLocationIdsListMaxSize  = 64;
+constexpr size_t kLocationDescriptorNameMaxSize = 128;
+
+constexpr size_t kDeviceDirectoryMaxSize   = 256;
+constexpr size_t kLocationDirectoryMaxSize = 64;
+
+class AttrAccess : public AttributeAccessInterface
+{
+public:
+    // Register for the EcosystemInformationCluster on all endpoints.
+    AttrAccess() : AttributeAccessInterface(Optional<EndpointId>::Missing(), Clusters::EcosystemInformation::Id) {}
+
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+};
+
+CHIP_ERROR AttrAccess::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    VerifyOrDie(aPath.mClusterId == Clusters::EcosystemInformation::Id);
+    switch (aPath.mAttributeId)
+    {
+    case Attributes::RemovedOn::Id:
+        return EcosystemInformationServer::Instance().EncodeRemovedOnAttribute(aPath.mEndpointId, aEncoder);
+    case Attributes::DeviceDirectory ::Id:
+        return EcosystemInformationServer::Instance().EncodeDeviceDirectoryAttribute(aPath.mEndpointId, aEncoder);
+    case Attributes::LocationDirectory ::Id:
+        return EcosystemInformationServer::Instance().EncodeLocationStructAttribute(aPath.mEndpointId, aEncoder);
+    default:
+        break;
+    }
+    return CHIP_NO_ERROR;
+}
+
+Structs::HomeLocationStruct::Type GetEncodableLocationDescriptorStruct(const LocationDescriptorStruct & aLocationDescriptor)
```

Comment:
This is a bit unsafe in that the caller's LocationDescriptorStruct could die while our return value is referencing it, but I don't really know a good way to address that in C++...

One thing you _could_ try to do to avoid temporaries being passed in is to have an overload that takes an rvalue ref and either fails to compile (would be easier if it were a deleted member method) or at least at runtime dies so that if something calls it we will notice.

Reply:
Yeah this was the best I could come up with. An earlier version had a failed direct encoder that was very ugly

I meant to add documentation to this to talk about the lifetime considerations (I will add comment before merging). I also chose to have this be in the anonymous namespace in the cpp file as it would would limit where and how it was used.

>One thing you could try to do to avoid temporaries being passed in is to have an overload that takes an rvalue ref and either fails to compile (would be easier if it were a deleted member method) or at least at runtime dies so that if something calls it we will notice.

Not sure I understand

Reply:
No, I am referring to overloading `GetEncodableLocationDescriptorStruct` with a version that takes a `LocationDescriptorStruct &&` which either causes compilation to fail when called or at least causes a ChipDie when called.

Reply:
Ahh.... understood. If this were accessible outside of this cpp file that might be worth it, but since this is used in the file with a warning label above it, and it is used properly here I am going to say it is not worth it.

I added a TODO comment so I might try the method of having this be a member method and deleting version that takes the `LocationDescriptorStruct &&`